index.bs

```diff
@@ -134,35 +134,29 @@ Framework {#framework}
 <p>To <dfn export id="sanitize-document-fragment">sanitize a document fragment</dfn> named <var>fragment</var> run these steps:
 
 <!-- Do we need to explicitly create a copy of the fragment? -->
-<ol>
-  <li><p>Initialize <var>node</var> to be the <var>fragment</var>'s root.
-  <li>
-    <p>while true
-    <ol>
-      <li>initialize *elementRemoved* to false.
-
-      <li>set <var>allowedElements</var> to the list of allowed elements for the node's `namespaceURI`.
-
-      <li>set <var>allowedAttributes</var> to the list of allowed attributes for the node's `namespaceURI`
-
-      <li>if the <var>node</var>'s local name is not in <var>allowedElements</var>:
-        <ol>
-        <li>Remove <var>node</var> and all of its ancestors from the tree and set <var>elementRemoved</var> to true
-        </ol>
-
-      <li>if <var>elementRemoved</var> is true, continue to next iteration.
-
-      <!-- node's local name is in the list of allowed element names -->
-      <li><p><a for=map>For each</a> <var>attribute</var> of <var>node</var>'s list of attributes
-        <ol>
-        <li>if <var>attribute</var> is not in <var>allowedAttributes</var>
-          <ol>
-            <li>Remove <var>attribute</var> from <var>node</var>'s attributes.
-          </ol>
-        </ol>
-    </ol>
-  <li>Return <var>fragment</var>.
-</ol>
+1. Initialize |node| to be the |fragment|'s root.
+1. While |node| is not null:
+  1. Initialize |elementShouldBeRemoved| to false.
+  1. Set |allowedElements| to the list of allowed elements for the node's `namespaceURI`.
+  1. Set |allowedAttributes| to the list of allowed attributes for the node's `namespaceURI`.
+  1. If the |node|'s local name is not in <var>allowedElements</var>:
+    1. Set |elementShouldBeRemoved| to true.
+
+  <!-- node's local name is in the list of allowed element names -->
+  1. [=map/For each=] |attribute| of |node|'s list of attributes:
+    1. If |attribute| is not in |allowedAttributes|.
+      1. Remove |attribute| from |node|'s attributes.
+
+  1. Initialize |current| to node.
+  1. Initialize |next| to null.
+  1. If |elementShouldBeRemoved| is false, set |next| to |next|'s [=tree/first child=].
+  1. While |next| is null and |current| is not null:
+    1. Set |next| to |next|'s [=tree/next sibling=]
+    1. Set |current| to *current.parentNode*.
+  1. If |elementShouldBeRemoved| is true:
+    1. [=/Remove=] |node| and all of its descendents from the tree.
+  1. Set |node| to |next|.
```

Comment:
I'm walking through this in my head, but it doesn't seem to work the way it needs to. Assume a structure like:

```
DocumentFragment
|
+--> P
     |
     +--> Text
     +--> Span
          |
          +--> Text
```

If we start at the `DocumentFragement` node, then as we enter, `node` is that node, and won't be removed. So we set `current` to the `DocumentFragment`, and `next` to null. Then we grab it's first child, and put it in `next`, so, at line 153, we have:

* `node` => `DocumentFragment`
* `current` => `DocumentFragment`
* `next` => `P`

That loop then then sets `next` to null (as `P` has no sibling), and sets `current` to `null` (as `DocumentFragment` has no parent).

158 then sets `node` to `null`, and we fall out of the loop.

I might be misunderstanding this, but I don't think it ends up descending the tree at all. Perhaps a recursive approach would be clearer?

As I noted in earlier comments, we might be able to avoid the complexity of defining this mechanism entirely. Sifting around a bit, it looks like some specs take an even simpler approach that we could follow, like https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree which just collects a list of all the elements and walks through the array rather than worrying about traversing a tree. That algorithm is admittedly a little more hand-wavey than I'd like, but it points in a direction I think might be reasonable!

Reply:
I totally agree this should be replaced with a reference to a properly specced algorithm somewhere, but I didn't find an obvious candidate. The problem with spec-ing this is mutating the tree while you walk through it. The CSS ref you give isn't well suited to this.

Unless we find a good match in another spec, we should IMHO proceed with this version (provided it is essentially correct, or at least more correct than the previous one). My main reason for looking at this at all is the sentence "Remove <var>node</var> and all of its ancestors from the tree [...]" in the current version. :)

-----

On the specific feedback: I agree up to the state for line 153. However, the loop condition is "next is null AND ...", and at this point next == P, so the loop never gets executed. The next iteration of the main loop will - as it should - then continue with node == P.

Reply:
Sorry I missed this response!

> I totally agree this should be replaced with a reference to a properly specced algorithm somewhere, but I didn't find an obvious candidate. The problem with spec-ing this is mutating the tree while you walk through it. The CSS ref you give isn't well suited to this.

I think you can get around the issues with mutating the tree as you go by walking from the bottom up. That is, the algorithm I linked grabs a list of elements in [shadow-including tree order](https://dom.spec.whatwg.org/#concept-shadow-including-tree-order), which DOM defines as a preorder, depth-first traversal of a tree. You're quite right that that's not appropriate as-is, as you might remove the root before traversing its children. That said, walking that list _backwards_ seems like it would work for you, as you'd encounter child nodes before their parents (see uses of "reverse tree order" in DOM and HTML), as would defining a new _postorder_, depth-first traversal type and walking it forwards.

If we take the first approach, we'd end up with an algorithm along the lines of:

```
1.   Let |nodes| be a list containing the [=inclusive descendants=] of |fragment|, in reverse [=tree order=].

2.  [=list/iterate|For each=] |node| in |nodes|:
    1.  // TODO: Add some filtering here to exclude non-Element nodes.
    2.  // TODO: Get the list of allowed elements and attribute for the node.
    3.  // TODO: Check the node and its attributes.
    4.  If |node| isn't as awesome as we'd like it to be, [=node/remove=] |node|.
```

I think that something like this could allow us to skip the complicated traversal logic entirely. It's arguably a little less efficient than walking from the top (as you might have removed a parent before processing both its children), but user agents would be free to implement more efficient mechanisms if necessary, so long as they had the same outcome. (Is there web-visible behavior wrapped up in removing elements from a `DocumentFragment`? Do mutation observers trigger, for instance? I don't know... we should probably check, as that would require more specific definitions here.)

All that said: I agree that this PR is better than the status quo. I think the approach I'm proposing here is clearer, but this is y'all's document: if you want to land it as-is, please do so. :)

> I agree up to the state for line 153. However, the loop condition is "next is null AND ...", and at this point next == P, so the loop never gets executed. The next iteration of the main loop will - as it should - then continue with node == P.

Thanks for this clarification, I did indeed miss the "and"!

Reply:
Ping?

Reply:
For a brief moment, I thought this was waiting for a 2nd review from myself. But we're waiting for @otherdaniel to continue here, right? :)

Reply:
I think so, yes. :)